src/Sema/arith.zig

```diff
@@ -0,0 +1,1609 @@
+    return .{ .overflow = res.overflow, .val = res.val };
+}
+
+/// `val` is of type `ty`.
+/// `ty` is a float, comptime_float, or vector thereof.
+pub fn negateFloat(
+    sema: *Sema,
+    ty: Type,
+    val: Value,
+) CompileError!Value {
+    const pt = sema.pt;
+    const zcu = pt.zcu;
+    if (val.isUndef(zcu)) return val;
+    switch (ty.zigTypeTag(zcu)) {
+        .vector => {
+            const scalar_ty = ty.childType(zcu);
+            const len = ty.vectorLen(zcu);
+            const result_elems = try sema.arena.alloc(InternPool.Index, len);
+            for (result_elems, 0..) |*result_elem, elem_idx| {
+                const elem = try val.elemValue(pt, elem_idx);
+                if (elem.isUndef(zcu)) {
+                    result_elem.* = elem.toIntern();
+                } else {
+                    result_elem.* = (try floatNeg(sema, elem, scalar_ty)).toIntern();
+                }
+            }
+            const result_val = try pt.intern(.{ .aggregate = .{
+                .ty = ty.toIntern(),
+                .storage = .{ .elems = result_elems },
+            } });
+            return .fromInterned(result_val);
+        },
+        .float, .comptime_float => return floatNeg(sema, val, ty),
+        else => unreachable,
+    }
+}
+
+/// Wraps on integers, but accepts floats.
+/// `lhs_val` and `rhs_val` are both of type `ty`.
+/// `ty` is an int, float, comptime_int, or comptime_float; *not* a vector.
+pub fn addMaybeWrap(
+    sema: *Sema,
+    ty: Type,
+    lhs: Value,
+    rhs: Value,
+) CompileError!Value {
+    const zcu = sema.pt.zcu;
+    if (lhs.isUndef(zcu)) return lhs;
+    if (lhs.isUndef(zcu)) return rhs;
+    switch (ty.zigTypeTag(zcu)) {
+        .int, .comptime_int => return (try intAddWithOverflow(sema, lhs, rhs, ty)).wrapped_result,
+        .float, .comptime_float => return floatAdd(sema, lhs, rhs, ty),
+        else => unreachable,
+    }
+}
+
+/// Wraps on integers, but accepts floats.
+/// `lhs_val` and `rhs_val` are both of type `ty`.
+/// `ty` is an int, float, comptime_int, or comptime_float; *not* a vector.
+pub fn subMaybeWrap(
+    sema: *Sema,
+    ty: Type,
+    lhs: Value,
+    rhs: Value,
+) CompileError!Value {
+    const zcu = sema.pt.zcu;
+    if (lhs.isUndef(zcu)) return lhs;
+    if (lhs.isUndef(zcu)) return rhs;
+    switch (ty.zigTypeTag(zcu)) {
+        .int, .comptime_int => return (try intSubWithOverflow(sema, lhs, rhs, ty)).wrapped_result,
+        .float, .comptime_float => return floatSub(sema, lhs, rhs, ty),
+        else => unreachable,
+    }
+}
+
+/// Wraps on integers, but accepts floats.
+/// `lhs_val` and `rhs_val` are both of type `ty`.
+/// `ty` is an int, float, comptime_int, or comptime_float; *not* a vector.
+pub fn mulMaybeWrap(
+    sema: *Sema,
+    ty: Type,
+    lhs: Value,
+    rhs: Value,
+) CompileError!Value {
+    const zcu = sema.pt.zcu;
+    if (lhs.isUndef(zcu)) return lhs;
+    if (lhs.isUndef(zcu)) return rhs;
```

Comment:
Is those supposed to be `rhs` on second line? @mlugg 

![image](https://github.com/user-attachments/assets/32a83a89-dfce-4441-a851-3d2f6e1c36f6)

Reply:
Oh, yes it is, thank you. I'll open a follow-up soon to fix that typo.